### ISSUE TITLE
sc-dtls-example-server performance benchmark

### ISSUE THREAD

**Issue by makareth** [eclipse/californium#351]
Hi there,
I'm currently benchmarking a slightly modified ExampleDTLSServer (as found in  californium/demo-apps/sc-dtls-example-server/src/main/java/org/eclipse/californium/scandium/examples).
My bench station is a 4Core/8Thread (Intel(R) Xeon(R) CPU E3-1270 v6 @ 3.80GHz), 32GB RAM machine running Ubuntu 16.04.2 LTS.

I find it difficult to handle more than 1000 requests per second. I'm trying to gather information regarding this performance. I expected the performance to be one to two orders of magnitude higher. I may be missing something, or messing with something.
Are there any benchmarking results published out there ? Maybe stress tools I could re-use to compare with the available benchmarks ? 

I'm trying to understand how the threading works, and how the decrypted payload is sent up to the application. From my understanding, Scandium creates a 6 * number_of_cpus thread pool. These threads are indeed created, but under a 1000 RPS load (at which UDP datagrams start being lost) I can only see one of them being busy (~25%), the other ones are almost idle (0% to 0.9%). What are those threads intended for ?

When the DTLS datagram is seen as an applicative DTLS datagram, it is decrypted and sent to the DTLSConnector instance which in turns handle it to the message handler (an instance of RawDataChannel). All of this seems to be done in the very same thread. The RawDataChannel advises to off-load the processing by threading it. Wouldn't it be suitable to have the messageHandler derive from a thread itself ?

But anyway, I'm eager to have an approximate of the expected performance of Scandium Sample DTLSServer, or any benchmark results.

Thank you

**Comment by sbernard31** [comment]
Which branch are you testing ?

**Comment by makareth** [comment]
@sbernard31 I cloned the master branch, commit 142d70ecb208a86ccbcf396a45f2047d05c53d9c.

**Comment by boaks** [comment]
Can you give us also informations about the clients you use?
- implementation also californium or something else?
- how many clients?

**Comment by boaks** [comment]
And also some informations on the kind of traffic you use:
- always new handshakes (so we talk about 1000 handshakes per second?)
- handshake followed by how many application messages
- how many application data is transfered

**Comment by makareth** [comment]
Regarding the test setup : 
- 3 load generation machines (8 cores, 32GB RAM)
- 1 stressed machine (as described previously)
The client we use is executed on each load generation node. Each node simulates a rampup from 1 to 500 clients, each of them sending 1000 frames. Each client sends one DTLS applicative payload per second.

We modified the sc-dtls-example-client as follow because the provided client only performs one application message transfer :
 
- Instanciate a ThreadPoolExecutor. Size is 500.
- Instantiate, add and submit 500 instances of a DTLSCLientRunnable to the ExecutorService, sleeping 1 second between each instantiation/add/submission. Hence spawning the 500 clients roughly takes 
 500seconds.
- The DTLSClientRunnable constructor instantates a DTLSConnector, calls setRawDataReceiver() and starts the DTLSConnector, instantiates InetSocketAddress().
- The DTLSClientRunnable run() method instantiates InetSocketAddress() (again?!) and starts a loop for 1000 iterations, with a a logging line and a Thread.sleep(1000), so that each client sends one frame every second.

So basically, the handshakes are received during the first 500 seconds of the test, during which it gets 1500 (500 per node) handshakes. More precisely, each second, the DTLS server receives 3 handshakes requests per second (1 client *per node* is spawned every second), for the first 500 seconds. Once a client has finished the handshake we start sending only applicative datagrams.
So, for one of the 1500 clients, we have 1 handshake (somewhere between t0 and t0+500seconds), followed by 1000 application messages.

From my observation, the application data size doesn't matter. We started with 800 bytes, but experience similar issue with the default "HELLO WORLD" payload.

**Comment by boaks** [comment]
Thanks! One more question: 
Do you know the "cpu's load" of the stressed machine?

**Comment by makareth** [comment]
You are welcome, thank you for your time at looking this !
Yes, the CPU load is steady around 5% once 1500 clients are concurrent and pushing application data. The LoadAverage-1min never exceeded 0.20.

**Comment by makareth** [comment]
A JProfiler overvew of the threads might help.
![screenshot from 2017-07-06 11-27-31](https://user-images.githubusercontent.com/2042576/27904798-3521ce9c-623e-11e7-91f1-d93c18739352.png)

**Comment by boaks** [comment]
> I find it difficult to handle more than 1000 requests per second.

1000 requests per second **per client**?

**Comment by boaks** [comment]
I tried to execute the benchmark also.

See PR #353

My results are, that locally execute with 30 clients I get about 

30000 messages in 7539 ms
3979 messages per s

May be you can make a smaller check using that client?

**Comment by makareth** [comment]
After raison kernel's rmem values we now top to 1500 overall requests per second, with 1500 clients (however each client sends 1 frame per second). We could spawn more client but would then experience higher loss rate.
Will try that client and see as it goes !

**Comment by boaks** [comment]
OK, so you increase the number of messages by more clients.
My test was using more messages by increasing the message frequence for some clients
(much more then 1 per second).
So I'm looking forward to your result.

**Comment by boaks** [comment]
Any new results?

**Comment by makareth** [comment]
We ran the suggested stress tool you suggested. We were able to achieve  as much as 4000 applicative datagrams per second with 30 concurrent clients. Both of us achieve comparable results.

However, I pinned down our "loss frames" issue to the socket.setReuseAddress(true) in DTLSConnector.java. I observed that during the test the load generator client occasionnally reused a previously setup and still in use port. AFAIK the is not an issue for the OS, however this could conofuse the DTLS server side. Actually, DTLS Server seems to see it as a renegotiation. From that moment, the datagrams sent by the original client using that port get a "MAC validation failed" on the server. using socket.setReuseAddress(false) seems to mitigate the issue. Why is the SO_REUSEADDR forced to true ?

I started to test again but can't make both your (and our derived-) tool spawn a decent number of clients. Looks like each DTLSConnector instanciates a thread pool. Thread count increases rapidly (48 threads per client) and the VM eventually fails creating native threads. I modified my configuration to allow from 12.000 to 24.000 threads but I really feel like this is an issue, and I wish I could get your thinking about this. I have 2 points I wish to discuss : 

1/ In my understanding, each DTLSConnector instance is created "per client" (even though public DTLSConnector(final DtlsConnectorConfig configuration, final SessionCache sessionCache) method makes me think that for each client, an InMemoryConnectionStore instance is created which looks odd to me). Shouldn't the InMemoryConnectionStore be shared among all DTLSConnector instances, instead of new InMemoryConnectionStore() in the DTLSConnector constructor ?

2/ Regarding the DTLSConnector again, it's start() method creates a StripedExecutorService. Pool size is, (for a 8thread machine) 48 threads. The formula for sizing is 6 * Runtime.getRuntime().availableProcessors(). However, since the number of clients can be high (thousands of them), and that the traffic generated per client could be low (this depends on the application throughput. In our case, the troughput is ~1KB/s, ~1 datagram per second) I come to the conclusion that such pool size is oversized for some applications. Too many threads (because point 1/), and minimal CPU load so no need for that many threads. Am I missing something ?

I maintain all this discussion here because the topic is mainly benchmarking. Please let me know if you want to move the discussion elsewhere or open a new issue.

**Comment by boaks** [comment]
Thanks for your feedback!

> Why is the SO_REUSEADDR forced to true ?

My impression is, that this was mainly introduced for restarting a connector, which is heavily used by the unit tests. Maybe we make this configurable.

> Looks like each DTLSConnector instanciates a thread pool. 
Yes, currently the implementation is focused on the server side ...

But I'm little mixed up by your feedback. Can clarify, which observation was done on the client side and which on the server?

**Comment by boaks** [comment]
> I maintain all this discussion here because the topic is mainly benchmarking. Please let me know if you want to move the discussion elsewhere or open a new issue.

For me it's OK to continue this issue ...

**Comment by makareth** [comment]
Regarding the SO_REUSEADDR, I agree that making it configurable (with default = false) would fit common cases.

The observation regarding the big number of threads was done client side. On the node I simulate 500 simultaneous clients I get tens of thousands threads created. All the remarks made in my previous comment are related to the client side.

On the server side, my overall number of threads is steady at 122 :) So clearly, the server side looks perfect to me.

**Comment by boaks** [comment]
I rebased the multi_client branch on the current 2.0.x head and add:
- address reuse configuration (default: true, but configured to disabled in the example client)
- setup the executor to use less threads in the example client.

Though rebased, the push was forced and so the pull may also require a force.

**Comment by makareth** [comment]
We just tested multi_client. We reached ~3600 messages per second this time (down from ~4000).

**Comment by makareth** [comment]
Back to 2.0.x branch, I made small changes so that when used in a "client" context, the DTLSConnector instantiates a small StripedExecutorService(2). When the DTLSConnector is in a "server" context, the DTLSConnector keeps the StripedExecutorService(DEFAULT_EXECUTOR_THREAD_POOL_SIZE).
Combined with the setReuseAddr(false), this allows for spawning multiple "client DTLSConnectors" (tested up to 1500).

So, now I am able to create many clients, I notice that on the server side, the DTLS-Receiver thread quickly reaches 100% CPU usage (so 1 core is fully used by that thread) as soon as there are ~1500 concurrent clients. It looks like we are now constrained by the fact that only 1 receiver thread handles the incoming packets into the server.

Would it make sense to increase the number of these DTLS-Receiver threads ?

**Comment by boaks** [comment]
I currently tested local on one machine with 500 clients (server and client using branch "multi_client").
My results (on Windows) are

500000 messages in 86365 ms
5789 messages per s

My CPU load was about 100%. So I reach more 3000.

> Back to 2.0.x branch, I made small changes

Do you use the changes in branch "multi_client"?

**Comment by boaks** [comment]
> Would it make sense to increase the number of these DTLS-Receiver threads ?

You may try it.

**Comment by boaks** [comment]
Just to mention:

the dtls example client extension is now merged into branch 2.0.x

**Comment by makareth** [comment]
Retested on 8 cores Ubuntu machines. I confirm the multi_client branch was tested. The results are shown below:
1 client : 609 msg/s
2 clients : 1447 msg/s
3 clients : 2749 msg/s
4 clients : 3122 msg/s
5 clients : 3657 msg/s
10 clients : 2095 msg/s
20 clients : 1555 msg/s
30 clients : 1548 msg/s
40 clients : 1440 msg/s
50 clients : 1327 msg/s

Not tested  beyond 50 clients. The load average was minimal during the whole test, would it be on the client or the server side.

I didn't dig the DTLS-Receiver since the profiling actually doesn't show nay evidence the bottleneck is there.

**Comment by boaks** [comment]
Any update?

Running a benchmark locally with disabled logging I get up to 20000 messages (server and clients on the same host).

**Comment by sbernard31** [comment]
On my side, sightly same result  (~27000 msg/s with 1000 clients/1000 messages, clients and server on the same host)
